File: tests/lib/helloworld.c
#include "nimscr.h"
#include <assert.h>
#include <stdint.h>
#include <stdio.h>

static struct nimscripter_Defines nimscripter_default_defines[2] = {
    {"nimscript", "true"}, {"nimconfig", "true"}};

void my_error_hook(char *filename, intptr_t line, intptr_t col, char *msg,
                   enum nimscripter_Severity sev) {
  printf("%s\n", msg);
}
void testImpl(struct nimscripter_VmArgs *args) {
  intptr_t val = nimscripter_vmargs_get_int(args, 0);
  printf("We got: %ld\n", val);
}

int main() {
  nimscripter_do_log = true;
  struct nimscripter_VmProcSignature testProc = {
      "testscript", "testInput", "hooks", testImpl};
  struct nimscripter_VmAddins addins = {&testProc, 1};
  nimscripter_errorHook = my_error_hook;

  char *searchPaths[] = {"tests/lib", "tests/lib/scriptdir"};
  struct nimscripter_WrappedInterpreter intr = nimscripter_load_script(
      "tests/lib/testscript.nims", &addins, searchPaths, 2,
      "/home/jason/.choosenim/toolchains/nim-2.0.2/lib",
      &nimscripter_default_defines[0], 2);

  struct nimscripter_opaque_PNode *ret = nimscripter_invoke(&intr, "doThing", 0, 0);
  intptr_t myVal = 0;
  nimscripter_pnode_get_int(ret, &myVal);

  printf("%ld\n", myVal);

  nimscripter_destroy_pnode(ret);

  struct nimscripter_opaque_PNode *input = nimscripter_int_node(500);

  assert(nimscripter_pnode_get_kind(ret) == nimscripter_nkIntLit);

  ret = nimscripter_invoke(&intr, "doOtherThing", &input, 1);

  char *myStr = "";
  nimscripter_pnode_get_string(ret, &myStr);
  printf("%s\n", myStr);

  assert(nimscripter_pnode_get_kind(ret) == nimscripter_nkStrLit);
  nimscripter_destroy_pnode(ret);
  nimscripter_destroy_pnode(input);

  input = nimscripter_new_node(nimscripter_nkBracket);

  for (int i = 0; i < 5; i++) {
    nimscripter_pnode_add(input, nimscripter_int_node(i));
  }

  ret = nimscripter_invoke(&intr, "arrTest", &input, 1);

  intptr_t passed = 0;
  assert(nimscripter_pnode_get_int(ret, &passed) && (bool)passed);
  nimscripter_destroy_pnode(ret);
  nimscripter_destroy_pnode(input);

  nimscripter_destroy_interpreter(&intr);
  return 0;
}

File: tests/lib/helloworld.nim
import ../../src/nimscripter/nimscr
import std/strformat

type Thing = object
  x, y: int
  s: set[char]

nimscr.init()
echo nimscr.version

errorHook = proc(name: cstring, line, col: int, msg: cstring, sev: Severity) {.cdecl.} =
  echo fmt"{line}:{col}; {msg}"

proc doThing(args: VmArgs) {.cdecl.} =
  assert args.getKind(0) == rkInt
  echo args.getInt(0)

let prc = 
  [
    VmProcSignature(package: "script", module: "script", name: "doSpecificThing", vmProc: doThing),
  ]


var addins = VmAddins(procs: cast[ptr UncheckedArray[typeof(prc[0])]](addr prc), procLen: prc.len)
const myScript = """
type Thing = object
  x, y: int
  s: set[char]

import std/json

proc doSpecificThing(i: int) = discard
proc getThing*(): Thing = Thing(x: 100, y: 200)

echo %*{"a": "Hello World"}

proc doThing*(): int =
  echo "Huh"
  30

proc doOtherThing*(a: int): string = $a

proc arrTest*(arr: openArray[int]): bool =
  echo arr
  arr == [0, 1, 2, 3, 4]

proc tupleTest*(a: int, b: string): (int, string) = (a, b)
proc inputTest*(a: (string, float, int, bool)) = echo a
doSpecificThing(200)
"""
writeFile("/tmp/script.nim", myScript)

let intr = loadScript(
  "/tmp/script.nim",
  addins,
  [],
  "/home/jason/.choosenim/toolchains/nim-#devel/lib"
)
var 
  ret = intr.invoke("doThing", [])
  myVal: BiggestInt

assert ret.kind == nkIntLit
assert ret.getInt myVal
echo myVal


ret = intr.invoke("doOtherThing", [newNode(500)])
assert ret.kind == nkStrLit
var str: cstring
assert ret.getString(str)
echo str

let input = newNode(nkBracket)
for i in 0..<5:
  input.add newNode(i)
ret = intr.invoke("arrTest", [input])

assert ret.getInt(myVal) and bool(myVal)

ret = intr.invoke("tupleTest", [newNode(100), newNode("hello")])
assert (int, string).fromVm(ret) == (100, "hello")
ret = intr.invoke("tupleTest", [newNode(3100), newNode("world")])
assert (int, string).fromVm(ret) == (3100, "world")
discard intr.invoke("inputTest", [newNode ("hello", 32f, 100, true)])
assert Thing.fromVm(intr.invoke("getThing", [])) == Thing(x: 100, y: 200)
